standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/IMetaStoreClient.java

```diff
@@ -733,6 +733,27 @@ Table getTable(String catName, String dbName, String tableName,
   List<Table> getTableObjectsByName(String dbName, List<String> tableNames)
       throws MetaException, InvalidOperationException, UnknownDBException, TException;
 
+
+  /**
+   * Get tables as objects (rather than just fetching their names).  This is more expensive and
+   * should only be used if you actually need all the information about the tables.
+   * @param request GetTablesRequest Object.
+   * @return A list of objects representing the tables.
+   *          Only the tables that can be retrieved from the database are returned.  For example,
+   *          if none of the requested tables could be retrieved, an empty list is returned.
+   *          There is no guarantee of ordering of the returned tables.
+   * @throws InvalidOperationException
+   *          The input to this operation is invalid (e.g., the list of tables names is null)
+   * @throws UnknownDBException
+   *          The requested database could not be fetched.
+   * @throws TException
+   *          A thrift communication error occurred
+   * @throws MetaException
+   *          Any other errors
+   */
+  GetTablesResult getTables(GetTablesRequest request)
```

Comment:
This method should have a different signature. So the Request/Result objects are serialized/deserialized between the client and the server. The client side application code (for example HiveServer2) should not have to deal with Request/Response objects.
So HS2 will invoke something like this
IMetastoreClient.getTableObjectsByName(String catName, String dbName, List<String> tableNames)
or
IMetastoreClient.getTableObjectsByName(String catName, String dbName, List<String> tableNames, GetProjectionSpec projectionSpec)

These 2 methods internally have to build a GetTablesRequest object based on the arguments, make the thrift call which returns the GetTablesResults object, which then have to be returned to calls as List<Table>

So the signature should be something like this
List<Table> getTables/getTableObjectsByName(catName, dbName, tableNames, projectionSpec)

Reply:
I agree to this.

Request/Response objects are thrift serialization objects and are not meant for use outside the HiveMetaStore (client and server).

I also agree to your argument that if more fields are added to them tomorrow, these will need to be accounted for across all its usages.

I have changed this and updated the pull request.